sdk/src/dlob/DLOB.ts

```diff
@@ -0,0 +1,863 @@
+			vBid,
+			slot,
+			marketType,
+			oraclePriceData
+		);
+
+		let nextAsk = askGenerator.next();
+		let nextBid = bidGenerator.next();
+
+		// First try to find orders that cross
+		while (!nextAsk.done && !nextBid.done) {
+			const { crossingNodes, exhaustedSide } = this.findCrossingOrders(
+				nextAsk.value,
+				nextBid.value,
+				oraclePriceData,
+				slot
+			);
+
+			const takerIsMaker =
+				crossingNodes?.makerNode !== undefined &&
+				crossingNodes.node.userAccount.equals(
+					crossingNodes.makerNode.userAccount
+				);
+
+			// Verify that each side is different user
+			if (crossingNodes && !takerIsMaker) {
+				nodesToFill.push(crossingNodes);
+			}
+
+			if (exhaustedSide === 'bid') {
+				nextBid = bidGenerator.next();
+			} else if (exhaustedSide === 'ask') {
+				nextAsk = askGenerator.next();
+			} else if (exhaustedSide === 'both') {
+				nextBid = bidGenerator.next();
+				nextAsk = askGenerator.next();
+			} else {
+				break;
+			}
+		}
+		return nodesToFill;
+	}
+
+	public findMarketNodesToFill(
+		marketIndex: BN,
+		slot: number,
+		marketType: MarketType
+	): NodeToFill[] {
+		const nodesToFill = new Array<NodeToFill>();
+		// Then see if there are orders to fill against vamm
+		for (const marketBid of this.getMarketBids(marketIndex, marketType)) {
+			if (isAuctionComplete(marketBid.order, slot)) {
+				nodesToFill.push({
+					node: marketBid,
+				});
+			}
+		}
+
+		for (const marketAsk of this.getMarketAsks(marketIndex, marketType)) {
+			if (isAuctionComplete(marketAsk.order, slot)) {
+				nodesToFill.push({
+					node: marketAsk,
+				});
+			}
+		}
+		return nodesToFill;
+	}
+
+	public findJitAuctionNodesToFill(
+		marketIndex: BN,
+		slot: number,
+		marketType: MarketType
+	): NodeToFill[] {
+		const nodesToFill = new Array<NodeToFill>();
+		// Then see if there are orders still in JIT auction
+		for (const marketBid of this.getMarketBids(marketIndex, marketType)) {
+			if (!isAuctionComplete(marketBid.order, slot)) {
+				nodesToFill.push({
+					node: marketBid,
+				});
+			}
+		}
+
+		for (const marketAsk of this.getMarketAsks(marketIndex, marketType)) {
+			if (!isAuctionComplete(marketAsk.order, slot)) {
+				nodesToFill.push({
+					node: marketAsk,
+				});
+			}
+		}
+		return nodesToFill;
+	}
+
+	public getMarketBids(
+		marketIndex: BN,
+		marketType: MarketType
+	): Generator<DLOBNode> {
+		const marketTypeStr = getVariant(marketType) as MarketTypeStr;
+		return this.orderLists
+			.get(marketTypeStr)
+			.get(marketIndex.toNumber())
+			.market.bid.getGenerator();
+	}
+
+	public getMarketAsks(
+		marketIndex: BN,
+		marketType: MarketType
+	): Generator<DLOBNode> {
+		const marketTypeStr = getVariant(marketType) as MarketTypeStr;
+		return this.orderLists
+			.get(marketTypeStr)
+			.get(marketIndex.toNumber())
+			.market.ask.getGenerator();
+	}
+
+	*getAsks(
+		marketIndex: BN,
+		vAsk: BN | undefined,
+		slot: number,
+		marketType: MarketType,
+		oraclePriceData?: OraclePriceData
```

Comment:
should we just require oracle to always be there? perp orders also need it to calculate the limit price after auction is complete